src/pymor/algorithms/timestepping.py

```diff
@@ -10,221 +10,602 @@
+                               np.array([1/8, 3/8, 3/8, 1/8])),
+            'RK4' :           (np.array([0, 1/2, 1/2, 1]),
+                               np.array([[0, 0, 0, 0], [1/2, 0, 0, 0], [0, 1/2, 0, 0], [0, 0, 1, 0]]),
+                               np.array([1/6, 1/3, 1/3, 1/6])),
+            }
+
+    def __init__(self, method, nt, num_values=None, interpolation='P1'):
+        super().__init__(num_values, interpolation)
+
+        assert isinstance(method, (tuple, str))
+        if isinstance(method, str):
+            assert method in self.available_RK_methods.keys()
+            self.butcher_array = (self.available_RK_methods[method])
+        else:
+            raise RuntimeError('Arbitrary butcher arrays not implemented yet!')
+
+        assert isinstance(nt, Number)
+        assert nt > 0
+
+        self.__auto_init(locals())
 
 
 def implicit_euler(A, F, M, U0, t0, t1, nt, mu=None, num_values=None, solver_options='operator'):
-    assert isinstance(A, Operator)
-    assert isinstance(F, (type(None), Operator, VectorArray))
-    assert isinstance(M, (type(None), Operator))
-    assert A.source == A.range
-    num_values = num_values or nt + 1
-    dt = (t1 - t0) / nt
-    DT = (t1 - t0) / (num_values - 1)
-
-    if F is None:
-        F_time_dep = False
-    elif isinstance(F, Operator):
-        assert F.source.dim == 1
-        assert F.range == A.range
-        F_time_dep = F.parametric and 't' in F.parameters
-        if not F_time_dep:
-            dt_F = F.as_vector(mu) * dt
-    else:
-        assert len(F) == 1
-        assert F in A.range
-        F_time_dep = False
-        dt_F = F * dt
-
-    if M is None:
-        from pymor.operators.constructions import IdentityOperator
-        M = IdentityOperator(A.source)
-
-    assert A.source == M.source == M.range
-    assert not M.parametric
-    assert U0 in A.source
-    assert len(U0) == 1
-
-    R = A.source.empty(reserve=nt+1)
-    R.append(U0)
-
-    options = (A.solver_options if solver_options == 'operator' else
-               M.solver_options if solver_options == 'mass' else
-               solver_options)
-    M_dt_A = (M + A * dt).with_(solver_options=options)
-    if not M_dt_A.parametric or 't' not in M_dt_A.parameters:
-        M_dt_A = M_dt_A.assemble(mu)
-
-    t = t0
-    U = U0.copy()
-
-    for n in range(nt):
-        t += dt
-        mu = mu.with_(t=t)
-        rhs = M.apply(U)
-        if F_time_dep:
-            dt_F = F.as_vector(mu) * dt
-        if F:
-            rhs += dt_F
-        U = M_dt_A.apply_inverse(rhs, mu=mu, initial_guess=U)
-        while t - t0 + (min(dt, DT) * 0.5) >= len(R) * DT:
-            R.append(U)
-
-    return R
+    time_stepper = ImplicitEulerTimeStepper(nt=nt, initial_time=t0, end_time=t1, num_values=num_values,
+                                            solver_options=solver_options, interpolation_order=0)
+    return time_stepper.solve(mu=mu)
 
 
 def explicit_euler(A, F, U0, t0, t1, nt, mu=None, num_values=None):
```

Comment:
Do we want to keep these methods or deprecate them? I think I would be in favor of deprecating ..

Reply:
Me too!

Reply:
done in 09931564f26a3e5e921d4591e26332337c1cb58a